File: scripts/Engineering/gui/engineering_diffraction/tabs/fitting/plotting/plot_model.py
# Mantid Repository : https://github.com/mantidproject/mantid
#
# Copyright &copy; 2020 ISIS Rutherford Appleton Laboratory UKRI,
#     NScD Oak Ridge National Laboratory, European Spallation Source
#     & Institut Laue - Langevin
# SPDX - License - Identifier: GPL - 3.0 +


class FittingPlotModel(object):
    def __init__(self):
        self.plotted_workspaces = []

    def get_plotted_workspaces(self):
        return self.plotted_workspaces

    def add_workspace_to_plot(self, ws, ax, plot_kwargs):
        ax.plot(ws, **plot_kwargs)
        self.plotted_workspaces.append(ws)

    def remove_workspace_from_plot(self, ws, ax):
        if ws in self.plotted_workspaces:
            self._remove_workspace_from_plot(ws, ax)
            self.plotted_workspaces.remove(ws)

    @staticmethod
    def _remove_workspace_from_plot(ws, ax):
        ax.remove_workspace_artists(ws)

    def remove_all_workspaces_from_plot(self, ax):
        for ws in self.plotted_workspaces:
            self._remove_workspace_from_plot(ws, ax)
        # Should already be empty, but just to be sure.
        self.plotted_workspaces.clear()

File: qt/widgets/plotting/src/PlotWidget/PlotModel.cpp
// Mantid Repository : https://github.com/mantidproject/mantid
//
// Copyright &copy; 2022 ISIS Rutherford Appleton Laboratory UKRI,
//   NScD Oak Ridge National Laboratory, European Spallation Source,
//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
// SPDX - License - Identifier: GPL - 3.0 +
#include "MantidQtWidgets/Plotting/PlotWidget/PlotModel.h"
#include "MantidAPI/MatrixWorkspace_fwd.h"

#include <vector>

using namespace Mantid::API;

namespace MantidQt::MantidWidgets {

std::vector<Mantid::API::MatrixWorkspace_sptr> PlotModel::getWorkspaces() const { return m_workspaces; }

std::vector<int> PlotModel::getWorkspaceIndices() const { return m_workspaceIndices; }

void PlotModel::clear() noexcept {
  m_workspaces.clear();
  m_workspaceIndices.clear();
}

void PlotModel::setSpectrum(const Mantid::API::MatrixWorkspace_sptr &ws, const size_t wsIndex) {
  m_workspaces = std::vector<MatrixWorkspace_sptr>{ws};
  m_workspaceIndices = std::vector<int>{static_cast<int>(wsIndex)};
}
} // namespace MantidQt::MantidWidgets
